pkg/signature/signer.go

```diff
@@ -0,0 +1,99 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signature
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+
+	// these ensure we have the implementations loaded
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+
+	"github.com/pkg/errors"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+
+	// these ensure we have the implementations loaded
+	_ "golang.org/x/crypto/sha3"
+)
+
+type Signer interface {
+	SignMessage(message []byte, opts ...SignerOption) ([]byte, error)
+	crypto.Signer
```

Comment:
I still firmly believe that our Signers shouldn't expose `crypto.Signer` directly, since there isn't a sane way to pass request context to implementations that require it

Reply:
I'm not sure I understand your objection on this point; in this PR, I leverage the fact that the third argument to `Sign(io.Reader, []byte, crypto.SignerOpts)` is an interface that we can implement and extend to pass the context into the call (done here in GCP KMS implementation: 
https://github.com/sigstore/sigstore/pull/69/files#diff-3a72d0bc4fa0dbf57003eab15443ab0c199be6dacb6c256fcd98609f60801335R159 )

If someone passes `signature.SignerOpts` which contained `WithContext(desiredCtx)`, then we can use it on the underlying external call to the KMS provider. If it's not provided by the caller of `Sign()` and we still need a context, then we could use the one set at the time the constructor was called.

Reply:
The issue is `crypto.Signer` includes `PublicKey`, which doesn't have any args:
```
Public() PublicKey
```

Reply:
IIRC one of the function examples that was the impetus for this change accepted a `crypto.Signer` but not any `crypto.SignerOpt`s, right? That would also break this scheme

Reply:
Another option might be:

```
type ContextSigner struct {
  signature.Signer
  ctx context.Context
}

func (c ContextSigner) Public() crypto.PublicKey {
  pk, err := c.PublicWithContext(WithContext(c.ctx))
  if err != nil {
    panic(err)
  }
  return pk
}
```